packages/core/src/main/ts/getDataProps.ts

```diff
@@ -0,0 +1,23 @@
+type DataProps<T> = {
+  [K in keyof T as K extends `data-${infer _U}` ? K : never]: T[K]
+}
+
+export const getDataProps = <T extends {}>(
+  props: T,
+): {
+  data: DataProps<T>
+  rest: Omit<T, keyof DataProps<T>>
+} => {
+  const data: Record<string, unknown> = {}
+  const rest: Record<string, unknown> = {}
+
+  for (const key of Object.keys(props)) {
+    if (key.startsWith('data-')) {
+      data[key] = props[key as keyof typeof props]
+    } else {
+      rest[key] = props[key as keyof typeof props]
+    }
+  }
+
+  return { data, rest } as any
```

Comment:
I'd expect that `getSmth` to return `smth` w/o any boxing. If we need the `rest` fieldset for some exotic case we still can mix it as a non-enumerable prop of `data`.